serve/triton/client.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2021 Zilliz. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import numpy as np
+import tritonclient.grpc as grpcclient
+import tritonclient.http as httpclient
+from tritonclient.utils import InferenceServerException
+from functools import partial
+
+model_name = 'pipeline'
+
+if sys.version_info >= (3, 0):
+	import queue
+else:
+	import Queue as queue
+
+class UserData:
+	def __init__(self):
+		self._completed_requests = queue.Queue()
+
+# Callback function used for async_stream_infer()
+def completion_callback(user_data, result, error):
+	# passing error raise and handling out
+	user_data._completed_requests.put((result, error))
+
+class Client():
+	def __init__(self, url, model_name=model_name, protocol='grpc'):
+		self._model_name = model_name
+		self.user_data = UserData()
+		if protocol == 'grpc':
+			self.protocol = 'grpc'
+			self.client = grpcclient.InferenceServerClient(url)
+			self.client.start_stream(partial(completion_callback, self.user_data))
+		else:
+			self.protocol = 'http'
+			self.client = httpclient.InferenceServerClient(url)
+
+	def __del__(self):
+		if self.protocol == 'grpc':
+			self.stop_stream()
+
+	def get_model_name(self):
+		return self._model_name
+
+	def async_set(self, iterator):
+		if self.protocol != 'grpc':
+			raise Exception('Streaming is only allowed with gRPC protocol')
+		inputs, outputs = self._solve_inputs_outputs(iterator)
+
+		sent_count = 0
+		responses = []
+		for i in range(len(iterator)):
+			sent_count += 1
+			if self.protocol == 'grpc':
+				self.client.async_infer(self.get_model_name(), inputs, partial(completion_callback, self.user_data), request_id=str(sent_count), outputs=outputs)
+		
+		processed_count = 0
+		while processed_count < sent_count:
+			(results, error) = self.user_data._completed_requests.get()
+			processed_count += 1
+			if error is not None:
+				sys.exit(1)
+			if processed_count <= len(iterator):
+				responses.append(results)
+		
+		return self._solve_responses(outputs, responses)
+
+	def stream(self, iterator):
```

Comment:
what is the difference between `async_set ` and `stream`, if the protocol is grpc, all of them will call `start_stream`